main/src/ca/uwaterloo/flix/api/lsp/provider/Visitors.scala

```diff
@@ -0,0 +1,316 @@
+        visitDef(visit, accept)(defn)
+      }
+    }}
+
+    root.effects.map{ case (_, v) => {
+      if (inside(accept, v)) {
+        visitEffect(visit, accept)(v)
+      }
+    }}
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+    // root.enums.map{ case (_, v) => visitEnum(???, ???)(v) };
+    // root.instances.map { case (_, v) => visitInstance(???, ???)(v) };
+    // root.modules.map { case (_, v) => visitModule(???, ???)(v) };
+    // root.restrictableEnums.map { case (_, v) => visitRestrictable(???, ???)(v) }; //experimental, prob should be removed
+    // root.sigs.map{ case (_, v) => visitSig(???, ???)(v) };
+    // root.structs.map{ case (_, v) => visitStruct(???, ???)(v) };
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+    // root.traits.map{ case (_, v) =>  visitTrait(???, ???)(v) };
+    // root.typeAliases.map{ case (_, v) => visitTypeAlias(???, ???)(v) };
+    // root.uses.map{ case (_, v) => visitUse(???, ???)(v) };
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val posLine = pos.line + 1
+    val posCol = pos.character + 1
+
+    // sp1 and sp2, by invariant, has the same source, so we can use either
+    (uri == loc.sp1.source.name) &&
+    (posLine >= loc.beginLine) &&
+    (posLine <= loc.endLine) &&
+    !(posLine == loc.beginLine && posCol < loc.beginCol) &&
+    !(posLine == loc.endLine && posCol > loc.endCol)
+  }
+
+  def inside(other: SourceLocation => Boolean, defn: Def): Boolean = {
+    other(defn.spec.loc) || other(defn.exp.loc) || other(defn.sym.loc)
+  }
+
+  def inside(other: SourceLocation => Boolean, expr: Expr): Boolean = other(expr.loc)
+
+  def inside(other: SourceLocation => Boolean, effect: Effect): Boolean = {
+    other(effect.loc) || effect.ann.annotations.exists(ann => inside(other, ann)) || other(effect.sym.loc)
+  }
+
+  def inside(other: SourceLocation => Boolean, ann: Annotation): Boolean = other(ann.loc)
+
+  def inside(loc1: SourceLocation, loc2: SourceLocation): Boolean = {
+    loc1.source == loc2.source &&
+    (loc1.beginLine >= loc2.beginLine) &&
+    (loc1.endLine <= loc2.endLine) &&
+    !(loc2.beginLine == loc1.beginLine && loc2.beginCol > loc1.beginCol) &&
+    !(loc1.endLine == loc2.endLine && loc1.endCol > loc2.endCol)
+  }
+
+  def visitEffect(visit: Entity => Unit, accept: SourceLocation => Boolean)(effect: Effect): Unit = {
+    visit(Entity.Effect(effect))
+    ???
+  }
+
+  def visitDef(visit: Entity => Unit, accept: SourceLocation => Boolean)(defn: Def): Unit = {
+    visit(Entity.Def(defn))
+    if (accept(defn.spec.loc)) {
+      ???
+    }
+
+    if (inside(accept, defn.exp)) {
+      visitExpr(visit, accept)(defn.exp)
+    }
+  }
+
+  def visitExpr(visit: Entity => Unit, accept: SourceLocation => Boolean)(expr: Expr): Unit = {
+    // TODO: handle mutually recursive calls to non expressions in expressions (fx annotations)
+    val recur = visitExpr(visit, accept)
```

Comment:
OK. I need to teach you something:

_first_ you write it in the obvious but long way.

_only then_ can you introduce the right abstractions and refactor.

You don't try to be smart from the beginning because you paint yourself into a corner because you do not yet fully understand the problem.

Reply:
So forget about this recurIf blah and just write the code in the obvious way with minimal fuss.

Only then can we discuss a better way. 

Do not immediately reach for: higher order functions, polymorphism, traits. Just plain functions and plain recursion.

Reply:
Wax on, wax off:

![CleanShot 2024-09-18 at 10 19 58@2x](https://github.com/user-attachments/assets/0da30a55-d494-4633-a3ff-d7b50ce27351)

Reply:
i see what you mean. I didn't think it was an issue here, since I considered the abstraction fairly simple. but maybe it's better to not use it anyways. I'll revert to the simpler way :)

Reply:
your point about painting myself into a corner, I think you're absolutely right. Thats the danger of premature abstraction. I think I do fall for that sometimes

Reply:
ive actually never watched karate kid. I think I will :p

Reply:
I've watched it now. I danced too much, I see that now *bows while maintaining eyecontact